src/index.ts

```diff
@@ -508,15 +520,64 @@ export class ServerlessClamscan extends Construct {
     }
   }
 
+  /**
+   * @returns ArnPrincipal the ARN of the assumed role principal for the scan function
+   */
+  get scanAssumedPrincipal(): ArnPrincipal {
+    if (this._scanFunction.role) {
+      const stack = Stack.of(this);
+      const scan_assumed_role = `arn:${stack.partition}:sts::${stack.account}:assumed-role/${this._scanFunction.role.roleName}/${this._scanFunction.functionName}`;
+      return new ArnPrincipal(scan_assumed_role);
+    } else {
+      throw new Error('The scan function role is undefined');
+    }
+  }
+
+
+  /**
+   * Returns the statement that should be added to the bucket policy
+     in order to prevent objects to be accessed when they are not clean
+     or there have been scanning errors: this policy should be added
+     manually if external buckets are passed to addSourceBucket()
+   * @param bucket The bucket which you need to protect with the policy
+   * @returns PolicyStatement the policy statement if available
+   */
+  getPolicyStatementForBucket(bucket: IBucket): PolicyStatement {
+    if (this._scanFunction.role) {
+      const scan_assumed_principal = this.scanAssumedPrincipal;
+      return new PolicyStatement({
+        effect: Effect.DENY,
+        actions: ['s3:GetObject'],
+        resources: [bucket.arnForObjects('*')],
+        notPrincipals: [this._scanFunction.role, scan_assumed_principal],
+        conditions: {
+          StringEquals: {
+            's3:ExistingObjectTag/scan-status': [
+              'IN PROGRESS',
+              'INFECTED',
+              'ERROR',
+            ],
+          },
+        },
+      });
+    } else {
+      throw new Error("Can't generate a valid S3 bucket policy, the scan function role is undefined");
+    }
+  }
+
   /**
    * Sets the specified S3 Bucket as a s3:ObjectCreate* for the ClamAV function.
      Grants the ClamAV function permissions to get and tag objects.
      Adds a bucket policy to disallow GetObject operations on files that are tagged 'IN PROGRESS', 'INFECTED', or 'ERROR'.
    * @param bucket The bucket to add the scanning bucket policy and s3:ObjectCreate* trigger to.
    */
-  addSourceBucket(bucket: Bucket) {
+  addSourceBucket(bucket: IBucket) {
+    if (!(bucket instanceof Bucket) && !this.useExternalBuckets) {
```

Comment:
Instead of casting we can use [addToResourcePolicy](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_s3.Bucket.html#addwbrtowbrresourcewbrpolicypermission) and [use the result](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_iam.AddToResourcePolicyResult.html) to throw an error if using an external Bucket and responsibility has not been accepted.

Reply:
From the doc:

> Note that the policy statement may or may not be added to the policy. For example, when an IBucket is created from an existing bucket, it's not possible to tell whether the bucket already has a policy attached, let alone to re-use that policy to add more statements to it. So it's safest to do nothing in these cases.

So it shouldn't do anything when using an existing bucket, I've added an `if` to check that the type is `Bucket` before calling it just as an additional (and possibly redundant) safeguard. I think it's best that the policy is not touched at all on existing buckets: the user can still retrieve the statement using the `getPolicyStatementFor(bucket)` method, and then decide what to do with it.

Reply:
ok, I think I get your point now: are you suggesting that since the method won't touch policies for existing buckets we can check the result and only raise the error if the policy wasn't added to an existing bucket and the user didn't set the accept responsibility flag? Sounds good to me as long as we're sure that policies for existing buckets are never touched by `addToResourcePolicy`.